.ci/README.md

```diff
@@ -1,150 +1,47 @@
-
-### Updating Containers
-Containers are stored in the `magic-modules` GCP project using GCR. Information on using GCR can be found here: https://cloud.google.com/container-registry/docs/pushing-and-pulling. Upon modifying an image please update the Dockerfile within the relevant sub-directory.
-
--------
-
-#### A side note on tracking submodules
-The merge job sets the submodules back to tracking their downstream repositories on `master`, by updating the submodule commits to point to the most recent commit on `master` for their respective downstream repositories.  This may not be ideal - other commits may have been made to `master` since the downstream PR was merged - however, this is the best way to ensure that we do not go backwards.  Imagine the following situation:
+## How the pipeline works
 
-`magic-modules` PR #1 is created.  `downstream-repo-a` PR #7 and `downstream-repo-b` PR #8 are created from that PR.  `magic-modules` PR #2 is created.  `downstream-repo-a` PR #9 is created from that PR.  PR #7 is merged, then PR #9.  Since all of PR #2's downstream PRs are merged, PR #2 is merged, and it includes the changes from PR #1.  PR #8 is finally merged, and so PR #1 is ready to be merged.
+### Generation & Diffing
+The generation / diff pipeline generates the downstream at the PR's merge commit and at the left-side parent (`HEAD~`), which guarantees isolation of exclusively the changes made by the PR in question.  It creates commits for the downstreams as before-and-after commits.  Those commits don't have relevant git history - they're not meant to be applied - but they are used in a [two-dot diff](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-comparing-branches-in-pull-requests#three-dot-and-two-dot-git-diff-comparisons).  This means that if there are no changes to the downstream, you'll see an empty diff.  These downstream branches are named `auto-pr-$number-old` and `auto-pr-$number` respectively.
 
-When merging PR #1, if we update `downstream-repo-a` to the merge commit for PR #7, we will go backwards, erasing PR #2.  If we update to `master`, we will definitely include both the changes of both PR #1 and PR #2.
+### Downstream Pushing
+The Magician maintains a set of tags called `$REPO-sync` that tracks the Magic Modules commit the downstreams are up to date with.
 
--------
+In effect, this means that each downstream commit will correspond 1:1 to an MM commit. If an MM commit had no changes in a downstream, no commit will be created.  We are enforcing squash-merges-only in Magic Modules.
 
-# Troubleshooting The Magician
+Downstream pushing can only take place if the sync tag points to the commit which precedes the commit that is being pushed.  It is possible for the pipeline to get stuck if a commit is merged into Magic Modules which cannot be generated - this happens most often when the Gemfile.lock is updated.  We expect the failure case to be somewhat uncommon. 
```

Comment:
Gemfiles have been updated twice in the last month. Do you think it's worth documenting what the Gemfile update should look like?

Reply:
Sure!